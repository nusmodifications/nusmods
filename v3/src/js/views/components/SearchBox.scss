@import "~styles/utils/modules-entry.scss";

.searchBox {
  position: sticky;
  top: $navbar-height;
  z-index: $module-finder-search-z-index;
  padding: 0 0 1rem;
  background: linear-gradient(to bottom, var(--body-bg) 3rem, var(--body-bg-transparent));
}

.searchWrapper {
  position: relative;

  input {
    padding: 0.8rem 1rem 0.8rem 2rem;
    border-width: 0 0 1px;
    border-radius: 0;

    &:focus {
      box-shadow: none;
    }
  }
}

.searchIcon {
  position: absolute;
  height: 100%;
  transition: color 0.3s;
}

.searchBoxFocused {
  .searchIcon {
    color: theme-color();
  }
}

.searchHelp {
  position: absolute;
  top: calc(50% - 0.9rem);
  right: 0.5rem;
  font-size: 0.9rem;
  color: var(--gray-light);
}

@include media-breakpoint-down(sm) {
  .searchBox {
    top: $navbar-height + $navtab-height;
    padding: 0 0 0.4rem;
    margin: -1rem (-$grid-gutter-width / 2) 0;
    border-top: 1px solid var(--gray-lighter);
    background: linear-gradient(to bottom, var(--body-bg) 2rem, rgba(#000, 0.5) 2rem, rgba(#000, 0));
  }

  .searchWrapper {
    input {
      padding: 0.8rem 0 0.8rem 2.2rem;
      font-size: 1rem;
      border: 0;
    }
  }

  .searchIcon {
    left: $grid-gutter-width / 2;
    width: 1rem;
    padding-bottom: 0.1rem; // Visual alignment
  }
}
