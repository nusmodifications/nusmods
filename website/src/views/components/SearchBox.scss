@import "~styles/utils/modules-entry.scss";

.searchWrapper {
  position: relative;

  input {
    padding: 0.8rem 1rem 0.8rem 2rem;
    border-width: 0 0 1px;
    border-radius: 0;

    &:focus {
      box-shadow: none;
    }
  }
}

.searchIcon {
  position: absolute;
  height: 100%;
  transition: color 0.3s;
}

.searchBoxFocused {
  .searchIcon {
    color: theme-color();
  }
}

.searchHelp {
  position: absolute;
  top: calc(50% - 0.9rem);
  right: 0.5rem;
  font-size: 0.9rem;
  color: var(--gray-light);
}

@include media-breakpoint-down(sm) {
  .searchWrapper {
    input {
      padding: 0.8rem 0 0.8rem 2.2rem;
      border: 0;
      font-size: 1rem;
    }
  }

  .searchIcon {
    left: $grid-gutter-width / 2;
    width: 1rem;
    padding-bottom: 0.1rem; // Visual alignment
  }
}

@include mobile-landscape {
  .searchBox {
    top: 0;
  }
}
