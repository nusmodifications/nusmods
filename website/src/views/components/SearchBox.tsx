import * as React from 'react';
import classnames from 'classnames';
import { debounce } from 'lodash';

import { Search } from 'views/components/icons';
import styles from './SearchBox.scss';

type Props = {
  className?: string;
  throttle: number;
  useInstantSearch: boolean;
  initialSearchTerm: string | null;
  placeholder: string;
  onSearch: (str: string) => void;
};

type State = {
  searchTerm: string;
  isFocused: boolean;
  hasChanges: boolean;
};

export default class SearchBox extends React.PureComponent<Props, State> {
  searchElement = React.createRef<HTMLInputElement>();

  constructor(props: Props) {
    super(props);

    this.state = {
      isFocused: false,
      searchTerm: this.props.initialSearchTerm || '',
      hasChanges: false,
    };

    this.props.onSearch(this.state.searchTerm);
  }

  onSubmit = () => {
    const element = this.searchElement.current;

    if (element) {
      const searchTerm = element.value;
      this.setState({ searchTerm });

      this.debouncedSearch(searchTerm);
      this.debouncedSearch.flush();
      element.blur();
    }
  };

  onInput = (evt: React.ChangeEvent<HTMLInputElement>) => {
    if (evt.target instanceof HTMLInputElement) {
      const searchTerm = evt.target.value;
      this.setState({ searchTerm, hasChanges: true });

      if (this.props.useInstantSearch) this.debouncedSearch(searchTerm);
    }
  };

  private search = (input: string) => {
    this.setState({ hasChanges: false });
    this.props.onSearch(input.trim());
  };

  // eslint-disable-next-line react/sort-comp
  private debouncedSearch = debounce(this.search, this.props.throttle, {
    leading: false,
  });

  showSubmitHelp() {
    return (
      !this.props.useInstantSearch &&
      this.state.hasChanges &&
      this.searchElement.current &&
      this.searchElement.current.value
    );
  }

  render() {
    return (
      <div
        className={classnames(this.props.className, {
          [styles.searchBoxFocused]: this.state.isFocused,
        })}
      >
        <label htmlFor="search-box" className="sr-only">
          Search
        </label>
        <form
          className={styles.searchWrapper}
          onSubmit={(evt) => {
            this.onSubmit();
            evt.preventDefault();
          }}
        >
          <Search className={styles.searchIcon} />
          <input
            id="search-box"
            className="form-control form-control-lg"
            type="search"
            autoComplete="off"
            ref={this.searchElement}
            value={this.state.searchTerm}
            onChange={this.onInput}
            onFocus={() => this.setState({ isFocused: true })}
            onBlur={() => {
              this.setState({ isFocused: false });
              this.onSubmit();
            }}
            placeholder={this.props.placeholder}
            spellCheck
          />
        </form>

        {this.showSubmitHelp() && <p className={styles.searchHelp}>Press enter to search</p>}
      </div>
    );
  }
}
