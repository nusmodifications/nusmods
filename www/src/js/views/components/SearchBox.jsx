// @flow
import React, { PureComponent } from 'react';
import classnames from 'classnames';
import { debounce } from 'lodash';

import { Search } from 'views/components/icons';
import styles from './SearchBox.scss';

type Props = {
  className?: string,
  throttle: number,
  useInstantSearch: boolean,
  initialSearchTerm: ?string,
  placeholder: string,
  onSearch: string => void,
};

type State = {
  searchTerm: string,
  isFocused: boolean,
  hasChanges: boolean,
};

export default class SearchBox extends PureComponent<Props, State> {
  props: Props;
  searchElement: ?HTMLInputElement;

  state: State = {
    isFocused: false,
    searchTerm: this.props.initialSearchTerm || '',
    hasChanges: false,
  };

  componentWillMount() {
    this.search(this.state.searchTerm);
  }

  onSubmit = () => {
    const { searchElement } = this;

    if (searchElement) {
      const searchTerm = searchElement.value;
      this.setState({ searchTerm });

      this.debouncedSearch(searchTerm);
      this.debouncedSearch.flush();
      searchElement.blur();
    }
  };

  onInput = (evt: Event) => {
    if (evt.target instanceof HTMLInputElement) {
      const searchTerm = evt.target.value;
      this.setState({ searchTerm, hasChanges: true });

      if (this.props.useInstantSearch) this.debouncedSearch(searchTerm);
    }
  };

  search = (input: string) => {
    this.setState({ hasChanges: false });
    this.props.onSearch(input.trim());
  };

  debouncedSearch: string => void = debounce(this.search, this.props.throttle, { leading: false });

  showSubmitHelp() {
    return (
      !this.props.useInstantSearch &&
      this.state.hasChanges &&
      this.searchElement &&
      this.searchElement.value
    );
  }

  render() {
    return (
      <div
        className={classnames(this.props.className, {
          [styles.searchBoxFocused]: this.state.isFocused,
        })}
      >
        <label htmlFor="search-box" className="sr-only">
          Search
        </label>
        <form
          className={styles.searchWrapper}
          onSubmit={(evt) => {
            this.onSubmit();
            evt.preventDefault();
          }}
        >
          <Search className={styles.searchIcon} />
          <input
            id="search-box"
            className="form-control form-control-lg"
            type="search"
            ref={(e) => {
              this.searchElement = e;
            }}
            value={this.state.searchTerm}
            onChange={this.onInput}
            onFocus={() => this.setState({ isFocused: true })}
            onBlur={() => {
              this.setState({ isFocused: false });
              this.onSubmit();
            }}
            placeholder={this.props.placeholder}
            spellCheck
          />
        </form>

        {this.showSubmitHelp() && <p className={styles.searchHelp}>Press enter to search</p>}
      </div>
    );
  }
}
