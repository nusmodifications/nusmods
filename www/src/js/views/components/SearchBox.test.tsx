// @flow

import React from 'react';
import { shallow } from 'enzyme';
import SearchBox from './SearchBox';

describe(SearchBox, () => {
  test('should match snapshot', () => {
    expect(
      shallow(
        <SearchBox
          throttle={0}
          useInstantSearch={false}
          initialSearchTerm=""
          placeholder=""
          onSearch={jest.fn()}
        />,
      ),
    ).toMatchSnapshot();

    expect(
      shallow(
        <SearchBox
          throttle={0}
          useInstantSearch={false}
          initialSearchTerm="Hello world"
          placeholder="Testing testing 123"
          onSearch={jest.fn()}
        />,
      ),
    ).toMatchSnapshot();
  });
});
