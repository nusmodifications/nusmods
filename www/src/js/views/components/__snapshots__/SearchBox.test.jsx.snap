// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchBox should match snapshot 1`] = `
<div
  className=""
>
  <label
    className="sr-only"
    htmlFor="search-box"
  >
    Search
  </label>
  <form
    className="searchWrapper"
    onSubmit={[Function]}
  >
    <Search
      className="searchIcon"
      color="currentColor"
      size="24"
    />
    <input
      autoComplete="off"
      className="form-control form-control-lg"
      id="search-box"
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      placeholder=""
      spellCheck={true}
      type="search"
      value=""
    />
  </form>
</div>
`;

exports[`SearchBox should match snapshot 2`] = `
<div
  className=""
>
  <label
    className="sr-only"
    htmlFor="search-box"
  >
    Search
  </label>
  <form
    className="searchWrapper"
    onSubmit={[Function]}
  >
    <Search
      className="searchIcon"
      color="currentColor"
      size="24"
    />
    <input
      autoComplete="off"
      className="form-control form-control-lg"
      id="search-box"
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      placeholder="Testing testing 123"
      spellCheck={true}
      type="search"
      value="Hello world"
    />
  </form>
</div>
`;
